### ISSUE TITLE
French translation

### ISSUE THREAD

**Issue by fred555547** [lenmus/lenmus#101]
I'm working on the french translation of lenmus. Is there a way to see the result of the translation with lenmus ?

**Comment by cecilios** [comment]
> I'm working on the french translation of lenmus

Glad to know. Thank you very much !!! . To see the results of your translation:

- For the GUI, first, you have to generate the lenmus_fr.mo file from the lenmus_fr.po file. Then replace the file lenmus_fr.mo with the new one in:
    Linux:    /usr/local/share/lenmus/5.6.1/locale/fr/lenmus_fr.mo
    Windows:  C:\Program Files\lenmus\locale\fr\lenmus_fr.mo

- For the eBooks it is more complex and I have nothing ready to be used by translators. So please send me the po file for any eBook at any stage and I will return you in short time the translated eBook. Yo have to replace the new eBook in:
    Linux:    /usr/local/share/lenmus/5.6.1/locale/fr/books/<new-translated-book>
    Windows:  C:\Program Files\lenmus\locale\fr\books\<new-translated-book>


The translation process for eBooks is not simple. The eBooks are written in English as XML files. See for instance, the TheoryHarmony book:
    https://github.com/lenmus/lenmus/tree/master/books/src/TheoryHarmony

The translation is done by using the PO files for each eBook:
1. The eBook sources in XML are processed by LenMus tool 'langtool' (https://github.com/lenmus/langtool).
2. 'langtool' program extracts the strings to translate and creates the PO file for the eBook.
3. The PO file is sent to the translator.
4. When the translated PO file is received back, the MO file is generated and the eBook is again processed by 'langtool' to generate the translated book.

Why to translate the eBooks using PO files instead of just translating the XML files? The point is that it is relatively simple to do a translation when you have to translate everything in a document. The difficult part comes when you have to maintain your work. Detecting which parts did change and need to be updated is tiring, error-prone and highly unpleasant.

Think about a typical review scenario for a translator:

1. Compare the previous English file with the new one.
2. Identify the changes.
3. Identify where theses changes must be introduced in the translated document.
4. Finally, do the real translation work.

Steps 1 to 3 are tedious and unpleasant. But by using PO files and 'langtool', the computer does these steps for you.


As said, send me at any time the po file for any eBook at any translation stage and I will return you as soon as possible the translated eBook to be tested.

Thank you!

**Comment by cecilios** [comment]
I didn't realize that I have not updated the _lenmus_fr.po_ file after the last release so the _lenmus_fr.po_ file in the repo do not contain the latest strings. But po files for eBooks are still valid as there were no changes to the eBooks. I have updated the _lenmus_fr.po_ file and it is now uploaded to the repo. Please download it from https://github.com/lenmus/lenmus/tree/master/locale/fr

If you already have started reviewing lenmus.po, please sent it to me. I will update it an return it to you so you don't lose the work done so far.

Thank you.

**Comment by fred555547** [comment]
Thank you Cecilios !
I've installed the v5.6.2 from Ubuntu snap : in the Gui, the translation was partial. You must notice that the files are in /usr/share/lenmus/...
Now, I have compiled and installed from the sources : lenmus takes my translation, except some entries are missing ... I double check with the link you provided and keep you informed.

**Comment by fred555547** [comment]
Hi Cecilios,
PoEdit cannot open the po files. The content of the po files is identical to the mo files ...

**Comment by fred555547** [comment]
Hi Cecilios,
I also tried to build the langtool tool but I got a fatal error :Building LenMus langtool
Scanning dependencies of target langtool
[ 18%] Building CXX object CMakeFiles/langtool.dir/src/command.cpp.o
[ 18%] Building CXX object CMakeFiles/langtool.dir/src/dlg_compile_book.cpp.o
[ 27%] Building CXX object CMakeFiles/langtool.dir/src/ebook_processor.cpp.o
[ 36%] Building CXX object CMakeFiles/langtool.dir/src/help_processor.cpp.o
In file included from /home/fred/project/langtool/src/ebook_processor.cpp:44:0:
/home/fred/project/langtool/packages/wxXml2/include/wx/xml2.h:38:10: fatal error: libxml/xmlmemory.h: Aucun fichier ou dossier de ce type
 #include <libxml/xmlmemory.h>
          ^~~~~~~~~~~~~~~~~~~~
compilation terminated.
CMakeFiles/langtool.dir/build.make:110: recipe for target 'CMakeFiles/langtool.dir/src/ebook_processor.cpp.o' failed
make[2]: *** [CMakeFiles/langtool.dir/src/ebook_processor.cpp.o] Error 1
make[2]: *** Attente des tâches non terminées....
In file included from /home/fred/project/langtool/src/help_processor.cpp:44:0:
/home/fred/project/langtool/packages/wxXml2/include/wx/xml2.h:38:10: fatal error: libxml/xmlmemory.h: Aucun fichier ou dossier de ce type
 #include <libxml/xmlmemory.h>
          ^~~~~~~~~~~~~~~~~~~~
compilation terminated.
CMakeFiles/langtool.dir/build.make:134: recipe for target 'CMakeFiles/langtool.dir/src/help_processor.cpp.o' failed
make[2]: *** [CMakeFiles/langtool.dir/src/help_processor.cpp.o] Error 1
CMakeFiles/Makefile2:67: recipe for target 'CMakeFiles/langtool.dir/all' failed
make[1]: *** [CMakeFiles/langtool.dir/all] Error 2
Makefile:83: recipe for target 'all' failed
make: *** [all] Error 2

**Comment by trebmuh** [comment]
Do you have the "_libxml2-dev_" package installed ?

**Comment by cecilios** [comment]
> PoEdit cannot open the po files. The content of the po files is identical to the mo files ...

Please download the PO file from here https://github.com/lenmus/lenmus/blob/master/locale/fr/lenmus_fr.po
You do not need to download the MO file. lenmus_fr.po is 161 KB and lenmus_fr.mo is 62,3 KB. They are different. PO files are plain text and MO files are just the same content but in binary and with indexes for quick access.

If you try to see the MO file at GitHub using the web browser, probably it will be downloaded and opened in your computer using poEdit (or the po editor you have) and it will look the same as if you has opened the po file. But they are different. You must not edit the MO files.

**Comment by cecilios** [comment]
As commented by @trebmuh you need to install the "libxml2-dev" package.

Langtool is a mess. At first, I used HTML, then XML, also some markup language. But I never cleaned the program for third party use. When you run the program you will see some pop up windows for debugging and trace. It is not ready for general use but it works.

**Comment by cecilios** [comment]
@fred555547 
> I've installed the v5.6.2 from Ubuntu snap

I have not created a snap package !!! 

> You must notice that the files are in /usr/share/lenmus/...

I've re-checked and it should be installed in /usr/local/share/lenmus/...  (!).

 Could you please give me the link to the snap page from which you downloaded LenMus?

**Comment by cecilios** [comment]
### How to use langtool to generate the translated eBook
When the PO file is translated, verify, generate the MO file and leave it in '_langtool/locale/xx_'

Now run langtool. You will see this screen:
![image](https://user-images.githubusercontent.com/5238679/79683905-b3ce4380-822d-11ea-80df-76fa33349c60.png)

Close the pop up. Then another pop-up appears:
![image](https://user-images.githubusercontent.com/5238679/79683919-da8c7a00-822d-11ea-9645-1bd11075ef36.png)

Close it or move it to left or right:
![image](https://user-images.githubusercontent.com/5238679/79683959-03147400-822e-11ea-89d7-31146d2dcced.png)

Now, choose menu 'eBook > Compile eBook'. The program ask for the source XML file and the desired languages. Chose French. An choose **the folder** containing the sources for the eBook to translate, for instance:
![image](https://user-images.githubusercontent.com/5238679/79684023-81711600-822e-11ea-848d-4a4db8f7bca3.png)

langtool will then translate the eBook, generate the compiled eBook file (.LMB) and leave it in folder 'lenmus\locale\xx'. When it finishes with the translation it display the last element translated and a pop-up window with the last string translated,  something like this:
![image](https://user-images.githubusercontent.com/5238679/79684044-b1201e00-822e-11ea-8a83-5d510fdd8fca.png)

If there are crashes the most probably cause is a string with un-paired tags in the PO file translation.

A mess!

**Comment by fred555547** [comment]
@cecilios 
> Do you have the "_libxml2-dev_" package installed ?
I'm using Xubuntu 18.04 LTS low latency
Yes I did ... But there might be an issue with the CMake configuration. I also installed CMake and performed with the following :
source code : /home/fred/project/langtool
binaries : /home/fred/project/langtool/z_bin
Generator for the project : Unix Makefiles
Use default native compilers
Configure :
/home/fred/project/langtool/z_bin
LIBXML2_LIBRARIES => /usr/lib/x86_64-linux-gnu/libxml2.so
wxWidgets => -L/usr/lib/x86_64-linux-gnu-pthread-lwx_gtk2u_aui-3.0-lwx_baseu-3.0-lwx_gtk2u_gl-3.0-lwx_baseu_net-3.0-lwx_gtk2u_richtext-3.0-lwx_baseu_xml-3.0-lwx_gtk2u_xrc-3.0-lwx_gtk2u_core-3.0-lwx_gtk2u_adv-3.0-lwx_gtk2u_html-3.0
SRC_DIR => /home/fred/project/langtool/src
WXXML2_DIR => /home/fred/project/langtool/packages/wxXml2
wxWidgets libraries => -L/usr/lib/x86_64-linux-gnu-pthread-lwx_gtk2u_aui-3.0-lwx_baseu-3.0-lwx_gtk2u_gl-3.0-lwx_baseu_net-3.0-lwx_gtk2u_richtext-3.0-lwx_baseu_xml-3.0-lwx_gtk2u_xrc-3.0-lwx_gtk2u_core-3.0-lwx_gtk2u_adv-3.0-lwx_gtk2u_html-3.0
LibXml2 libraries => /usr/lib/x86_64-linux-gnu/libxml2.so
Configuring done
Generating done
And then, as I'm not a geek, I don't known how to proceed ...

I you have some ideas.

**Comment by fred555547** [comment]
@cecilios  : I grabbed the po file with wget but the content is a mo file.
wget https://github.com/lenmus/lenmus/blob/master/locale/fr/lenmus_fr.po
But I got it in raw format and firefox ;-)



> > I've installed the v5.6.2 from Ubuntu snap
> 
> I have not created a snap package !!!
I'm sorry : I downloaded the deb file and the installation was done by "Logiciel", and Ubuntu tool.

**Comment by cecilios** [comment]
To build, the simples approach is to open a command window and do:
```
cd <my/projects/folder>/langtool
mkdir mybuild
cd mybuild
cmake -DLIBXML2_LIBRARIES=/usr/lib/x86_64-linux-gnu/libxml2.so.2 ..
make
```
Change `/usr/lib/x86_64-linux-gnu/libxml2.so.2` by the path to your libxml2.so library. Do not forget the two dots at the end of the cmake command, for pointing to parent folder.

This will create the executable at ../z_build. To run it:
```
cd ../z_build
./langtool
```

**Comment by cecilios** [comment]
If you do
```
wget https://github.com/lenmus/lenmus/blob/master/locale/fr/lenmus_fr.po
```

This will  download the HTML page  not the file!

To download the file you need to download the raw content (what you see when clicking on button [Raw] in previous page). Do
```
wget https://raw.githubusercontent.com/lenmus/lenmus/master/locale/fr/lenmus_fr.po
```

**Comment by fred555547** [comment]
I did the following 
`cmake -DLIBXML2_LIBRARIES=/usr/lib/x86_64-linux-gnu/libxml2.so.2 -DLIBXML2_INCLUDE_DIR=/usr/include/libxml2/libxml ..`
I always get the same error
`
-- LIBXML2_LIBRARIES => /usr/lib/x86_64-linux-gnu/libxml2.so.2
-- wxWidgets => -L/usr/lib/x86_64-linux-gnu-pthread-lwx_gtk2u_aui-3.0-lwx_baseu-3.0-lwx_gtk2u_gl-3.0-lwx_baseu_net-3.0-lwx_gtk2u_richtext-3.0-lwx_baseu_xml-3.0-lwx_gtk2u_xrc-3.0-lwx_gtk2u_core-3.0-lwx_gtk2u_adv-3.0-lwx_gtk2u_html-3.0
-- SRC_DIR => /home/fred/project/langtool/src
-- WXXML2_DIR => /home/fred/project/langtool/packages/wxXml2
-- wxWidgets libraries => -L/usr/lib/x86_64-linux-gnu-pthread-lwx_gtk2u_aui-3.0-lwx_baseu-3.0-lwx_gtk2u_gl-3.0-lwx_baseu_net-3.0-lwx_gtk2u_richtext-3.0-lwx_baseu_xml-3.0-lwx_gtk2u_xrc-3.0-lwx_gtk2u_core-3.0-lwx_gtk2u_adv-3.0-lwx_gtk2u_html-3.0
-- LibXml2 libraries => /usr/lib/x86_64-linux-gnu/libxml2.so.2
-- Configuring done
-- Generating done
-- Build files have been written to: /home/fred/project/langtool/mybuild
fred@fred:~/project/langtool/mybuild$ make
[  9%] Building CXX object CMakeFiles/langtool.dir/src/ebook_processor.cpp.o
In file included from /home/fred/project/langtool/src/ebook_processor.cpp:44:0:
/home/fred/project/langtool/packages/wxXml2/include/wx/xml2.h:38:10: fatal error: libxml/xmlmemory.h: Aucun fichier ou dossier de ce type
 #include <libxml/xmlmemory.h>
          ^~~~~~~~~~~~~~~~~~~~
compilation terminated.
CMakeFiles/langtool.dir/build.make:110: recipe for target 'CMakeFiles/langtool.dir/src/ebook_processor.cpp.o' failed
make[2]: *** [CMakeFiles/langtool.dir/src/ebook_processor.cpp.o] Error 1
CMakeFiles/Makefile2:67: recipe for target 'CMakeFiles/langtool.dir/all' failed
make[1]: *** [CMakeFiles/langtool.dir/all] Error 2
Makefile:83: recipe for target 'all' failed
make: *** [all] Error 2

`

**Comment by fred555547** [comment]
It seems there is an issue with the path /usr/include/libxml2.
I attached CmakeCache.txt that seems OK.

I when a little bit further when giving the full path of xmlmemory.h in ~/project/langtool/packages/wxXml2/include/wx/xml2.h
`#include <libxml/xmlmemory.h>` 
replaced by
`#include </usr/include/libxml2/libxml/xmlmemory.h>` 
Next error was with the file
xmlversion.h


[CMakeCache.txt](https://github.com/lenmus/lenmus/files/4499055/CMakeCache.txt)

**Comment by cecilios** [comment]
In my computer I do not have libxml2-dev installed, only libxml2. After investigating how can I build without libxml2-dev (the headers) the answer is that langtool uses wxXml2, a wxWidgets wrapper for libxml2. And the sources include the headers for libxml2, that I did'nt uploaded to the repo.
It is not clear if wxXml2 is currently maintainded.

I have to review the build process for langtool. There are several things to do:
- Fix the cmake module for finding libxml2
- Check to build using libxml2-dev package
- Check the maintenace state of wxXml2 or, better, replace it to have less dependencies.

As this will take me some time, I have uploaded to the repo the libxml2 headers that wxWxml2 uses (works in my computer). Update your local copy of langtool. Then try to build with:
```
cd /home/fred/project/langtool
mkdir mybuild
cd mybuild
cmake -DLIBXML2_LIBRARIES=/usr/lib/x86_64-linux-gnu/libxml2.so ..
make
```
You don't need to specify -DLIBXML2_INCLUDE_DIR=/usr/include/libxml2/libxml

Tell me this still fails.
Remember that before trying a new build you have to clear the build forder:
```
cd /home/fred/project/langtool/mybuild
rm -rf *        # be sure your are in the right folder before doing this !!!!
cmake [.......]
```

**Comment by cecilios** [comment]
Ok, now CMakeList.txt is fixed. You no longer need the headers I uploaded a few minutes agoo. I have removed them. Download the latest version of CMakeList.txt and try again:
```
cd /home/fred/project/langtool
mkdir mybuild
cd mybuild
cmake ..
make
```
You don't need to specify anything in cmake command. Please tell me if this still fails.
Remember that before trying a new build you have to clear the build forder:
```
cd /home/fred/project/langtool/mybuild
rm -rf *        # be sure your are in the right folder before doing this !!!!
cmake ..
make
```

**Comment by cecilios** [comment]
Ah! and you need to install libxml2-dev:
```
sudo apt-get install libxml2-dev
```

**Comment by fred555547** [comment]
Hi,
I fixed the problem. In theCMakeLists.txt file, I added the line ` /usr/include/libxml2` in the `include_directories` section.
And now it's OK and I'm happy
[CMakeLists.txt](https://github.com/lenmus/lenmus/files/4499235/CMakeLists.txt)

**Comment by cecilios** [comment]
:thumbsup: Yes, that was the solution. But the right way of including the path is not as some constant path because then it only works in your platform. You have to include the path found when looking for libxml2. Look at how I did it:
https://github.com/lenmus/langtool/commit/bd6857a86ecbf93df1e588e09c809fd6b44958e3

**Comment by fred555547** [comment]
This the first time I fix a bug ! Anyway, your way is the best, for sure.
Now, I'm going the improve the french translation, using langtool. The reason is that, as in spannish, adverbs have a gender. For exemple augmented can translated to 'augmenté' when dealing with 'un intervalle augmenté' or augmentée if dealing with 'une quarte aumentée '.

**Comment by cecilios** [comment]
That is going to be difficult to fix for now. The only solution I can think to solve this is to not split sentences. That is, translate full sentences: "augmented second", "augmented third", etc. instead of a list "second", "third" ... and a few adjectives "augmented", "diminished", etc. than later are combined in the program. This is going to be more work for translators but the translation will be better. I'm afraid this can not be done in short time as this change requires to review the source code to locate all strings and make the appropriate changes.

I have opened issue #105 for this.

**Comment by fred555547** [comment]
There is another way : the translation of the english eBooks ... Anayway, let's go in the usual way, we will see if we need to improve it later. This gives me time to see and analyse the résult

**Comment by cecilios** [comment]
> There is another way : the translation of the english eBooks

I have doubts, as for eBooks the translation units should be full sentences. I see the problem in the exercises (lenmus_xx.po file) not in the texts for the eBooks.

**Comment by fred555547** [comment]
I founded some spanish sentences in the lenmus_fr.po.
`Acorde completo. Contiene todas las notas (en todo caso, elidir la 5ª)`
what is the meaning of 'completo' and 'elidir la 5ª'

**Comment by cecilios** [comment]
Sorry. If you find any other one in Spanish (I don't think so) forget about those sentences. Those are messages not yet used that I didn't know how to translate into English. The approximate meaning is "Full chord. Contains all the notes (in any case, supress the 5th)" but I have doubts about English musical terms, such as Spanish _'elidir'_ (suppress?, elide?), or _'acorde completo'_ (full chord? complete chord?)

**Comment by fred555547** [comment]
This is interesting. Without the context (I have to read the exercise book I guess), a full chord is a triad : dominant, major third and fifth. So if you remove the fifth ... I might be wrong ...
The best is to give an exemple so I can compare to french. For some translations, I use the lilypond glossary.
It seems the term Reload, in the File menu, is not in the lenmus_fr.po. Could you confirm ?

**Comment by cecilios** [comment]
> a full chord is a triad : dominant, major third and fifth. So if you remove the fifth ... 

Yes, that is the meaning of _Acorde completo_. But I don't know how is it named in English, probably _full chord_ but I need to confirm. An in certain cases, in harmony, the fifth of the chord is removed. That is a chord with _la 5ª elidida" (in English, supressed? removed? elided?). But that string is not necessary. Is part on some exercises for studying chord progressions that are not yet finished. Forget about that translation.

> It seems the term Reload, in the File menu, is not in the lenmus_fr.po. Could you confirm ?

I've been checking and ... good catch!. In the code, the word "Reload" (used in the File menu) is not marked for translation and, thus, it is not included in the PO file. I have fixed the code but, in any case, it is impossible to translate for current binary packages. Once your translation is ready and I receive your PO file I will update it and fix the translation. What is the French translation for it, _Recharger_?

**Comment by fred555547** [comment]
Perfecto ! Eso es la palabra en francés (sorry I've a french keyboard)

**Comment by fred555547** [comment]
Hi Cecilios,
I wanted to test L1_MusicReading_fr.po. So I used LangTool with the following configuration :
`Preparing to process eMusicBook. src='/home/fred/project/lenmus/books/src/L1_MusicReading/L1_MusicReading.xml', dest='/home/fred/project/lenmus/locale/fr/books'
Loading file /home/fred/project/lenmus/books/src/L1_MusicReading/L1_MusicReading.xml`

Compilation is OK : 
`
Compilation done successfully
`
Then I copied the lmb file
`
root@fred:/usr/local/share/lenmus/5.6.2/locale/fr/books# cp /home/fred/project/lenmus/locale/fr/books/L1_MusicReading.lmb .
`
Now LenMus crashes when I select the ebook L1_MusicReading
Note : inthe folder `home/fred/project/langtool/locale/fr` there is only the file `L1_MusicReading_fr.mo`
[L1_MusicReading_fr.mo.tar.gz](https://github.com/lenmus/lenmus/files/4532956/L1_MusicReading_fr.mo.tar.gz)

**Comment by fred555547** [comment]
The error message is:
`Abandon (core dumped)
`

**Comment by fred555547** [comment]
A long time ago, I learned spanish. It could be easier for me to translate from spanish to french , eventually with the support of english ...

**Comment by cecilios** [comment]
The problem with translated books is crashes. When doing a translation, the translator some times makes syntactic errors with tags included in the translation. e.g. `This <a1>sentence</a1> is transladed` is translated as `Esta <a1>frase<a1> está traducida` or `Esta <a1>frase está traducida`. poEdit does not detect this. And langtool and lenmus might crash. So, when I receive a translation my first task is running lenmust and if it crashes, I run the debugger to identify the bad string and to fix it in the translation.

**Comment by cecilios** [comment]
>  It could be easier for me to translate from spanish to french

That is always possible and simple: Take the lenmus_es.po file, rename it as lenmus_fr.po and translate it. You will see both, the English sentence and the Spanish translation. Then, just replace the Spanish translation with your French translation.

**Comment by fred555547** [comment]
To get out of the crash, I reinstalled LenMus (following the procedure on the website). It's OK when run it, except that for the L1-MusicReading I get the 'old' fr translation, not my new one. Could you tell me how to proceed so I see 'my' translation in LenMus ?

**Comment by cecilios** [comment]
https://github.com/lenmus/lenmus/issues/101#issuecomment-615212095

> For the eBooks ... you have to replace the new eBook in:
Linux: /usr/local/share/lenmus/5.6.1/locale/fr/books/

**Comment by fred555547** [comment]
A long time ago, I learned spanish. It could be easier for me to translate from spanish to french , eventually with the support of english ...

**Comment by fred555547** [comment]
Hi Cecilios,
always the same guy ...
On the screenshot you can see that the translation applies only on the first occurrences for the staff lines and the staff spaces. I don't know how to handle this
![Capture d’écran_2020-04-25_23-52-03](https://user-images.githubusercontent.com/6939843/80291724-603d8780-8750-11ea-8127-b867d42f9c66.png)

**Comment by cecilios** [comment]
I've been checking and there is a bug in langtool causing that only the first string in an score was added to the translation. I have fixed the langtool program and have uploaded the updated POT files.

For updating your L1_MusicReading_fr.po file without losing your already translated strings you need to download the latest L1_MusicReading.pot file from:
https://github.com/lenmus/langtool/tree/master/locale/pot-files

Then open your L1_MusicReading_fr.po with poEdit, go to main menu and chose: Catalog > Update from POT file ...
Then select the downloaded POT file.

**Comment by fred555547** [comment]
Due to a problem on my PC, I had to re-install linux (Xubuntu). I have and error with the make command:
`/home/fred/project/lenmus/lomse/trunk/src/platform/lomse_linux.cpp:37:10: fatal error: fontconfig.h: Aucun fichier ou dossier de ce type
 #include <fontconfig.h>     //to use fontconfig
          ^~~~~~~~~~~~~~
compilation terminated.
CMakeFiles/lenmus.dir/build.make:3553: recipe for target 'CMakeFiles/lenmus.dir/lomse/trunk/src/platform/lomse_linux.cpp.o' failed
make[2]: *** [CMakeFiles/lenmus.dir/lomse/trunk/src/platform/lomse_linux.cpp.o] Error 1
CMakeFiles/Makefile2:152: recipe for target 'CMakeFiles/lenmus.dir/all' failed
make[1]: *** [CMakeFiles/lenmus.dir/all] Error 2
Makefile:171: recipe for target 'all' failed
make: *** [all] Error 2
`
Did you modify something ?

**Comment by fred555547** [comment]
Hi Cecilios,
Oups ! I fixed it with the following
`sudo apt-get install libfontconfig1-dev`
Maybe should add it to the installation manual 
`http://www.lenmus.org/en/phonascus/building-linux`

**Comment by cecilios** [comment]
Thank you for reporting. I've updated the build instructions.

**Comment by fred555547** [comment]
Hi Cecilios,
Just a question about the "General Exercises". In the options of the intervals' comparison and intervals' comparison, there is a tab for the Accidentals. Is it normal to select more than one  key signature. In the Accidental part, the fist item says "Use only notes from the key signature" ?

**Comment by cecilios** [comment]
When trying to develop absolute ear on a key you would need intervals formed exclusively by notes in the key signature. My understanding is that if you choose the option _"Use only notes from the key signature"_ you normally would select only one key signature, but this is open to the user.

LenMus allows to set any possible option for the intervals to generate, so that the user can set the exercises according to what he/she needs to practice. The drawback of this approach is that it is the user responsibility to select a coherent set of settings. Currently we are working on defining sets of coherent predefined settings for the exercises, graded by levels of difficulty/knowledge, to avoid the trouble for the user of having to decide the settings. Nevertheless, the option for free customization will still be available. But this is for the next release of lenmus.

**Comment by fred555547** [comment]
This is OK. In range note selector, the name of the notes is not translated.

**Comment by cecilios** [comment]
Thanks for reporting. I've checked the code and this is something that need to be fixed. Today I cannot do it. I open issue #107 and will fix it as soon as possible.

**Comment by fred555547** [comment]
With the following version
`Version du programme : 5.6.2+fa62d8c [5.6.2]
Date de création du programme : May 2 2020, 23:55:23
wxWidgets 3.0.4
lomse 0.27.0-nogit [0.27.0]
wxMidi 2.0
sqlite3 3.22.0
Informations sur votre ordinateur :
Charset encoding: UTF-8`
LenMus crashes when opening by "File/Open Book..." and select  'Study guide' link in the window.
Another point (esthetic) : is it possible to modify the order of the available books, i.e. level 1, level2, harmony and exercices ?

**Comment by fred555547** [comment]
When opening a book, it's name is in english in the tab (GeneralExercises, L1_MusicReading ...)

**Comment by fred555547** [comment]
Hi Cecilios,
here is my initial contribution to the french translation, with:
- an update of study-guide, 
- a partial translation of wxwidgets_fr.po and mo, 
- lenmus_fr.po and mo,
- L1_MusicReading book,
- GeneralExrecices book,
- and some comments and questions (in english and spanish ... ) in the file Modifications.txt. These are propositions. You'll see them using the french configuration.
For the books, I did it the dirty way : a direct translation in the lmd files, including the toc from the spanish wording, that is without using langtool. I'll use it later to check my work.

[lenmus_fr_fba.zip](https://github.com/lenmus/lenmus/files/4584161/lenmus_fr_fba.zip)

**Comment by cecilios** [comment]
> LenMus crashes when opening by "File/Open Book..." and select 'Study guide' link in the window.

Thank you. Bug confirmed. I open issue #108.

**Comment by cecilios** [comment]
> is it possible to modify the order of the available books, i.e. level 1, level2, harmony and exercices ?

No, currently it is not possible change the order. For next version, we are redesigning the main window and the books will be ordered by levels. Exercises will not be, probabliy, a book but a link in another place.

**Comment by cecilios** [comment]
> here is my initial contribution to the french translation

Thank you very much. I will review all this and tell you something in a few days. These days I'm very bussy and don't have time for LenMus. Sorry!

> a partial translation of wxwidgets_fr.po

Could you please submit this to wxWidgets following the instructions at https://www.wxwidgets.org/about/translations/ ? I will use your wxwidgets_fr.po but it will become obsolete in future as, from time to time, I download the lates wxWidgets PO files.

**Comment by cecilios** [comment]
Merci beaucoup pour l'effort fait pour réviser la traduction française. Quand j'ai an enfant, j'ai étudié français et je le parlais. Mais avec le temps, je avais seulment utilisais l'anglais et oujurd'hui j'ai oublié le français. Je le comprendre un peu mais je ne peux pas le parler, encore moins d'écrire en français. Sorry!

I you agree, I would like to add your name in all credits (program, website, etc.) How should I include your name? As "fred555547"? Or do you prefer to use your real name? (I do not have it).

After analyzing what you send me:
- I will replace current study-guide, lenmus_fr.po and lenmus_fr.mo with your updated files.
- I also will replace books L1_MusicReading and GeneralExrecices, but be aware that your translations will be lost as soon as I have to make any change, because they will be re-generated from English version and PO files. I will try to not modify your new translated books but, when inadvertently using the scripts I have, they will process all eBooks and your translation will be lost. Please, consider updating the eBooks PO files with your translation. As we are starting to prepare new books, under the direction of group of professors from the Department of Music Pedagogy of a Spanish University, perhaps it would be better not to spend much time in current eBooks and collaborate instead in the translation of the new eBooks in preparation. We have just started, so the task to do will be to translate as we develop the material. I expect that in average, this will be no more than 30 strings per month.

As to the wxWidgets_fr.po file, comparing changes with the current one in LenMus, I observe some strange things. I use the Meld program (https://meldmerge.org/) (install it with `sudo apt-get install meld`) but you don't need to install it, just open the PO file with a text editor:
- It contains many sentences in Spanish that in the original were in French. For example, look at lines 37, 126 or 221.
- The spaces are important in some texts and spaces have been removed in places where they are needed. For example, lines 59 and following.
- And it does not seem that there are sentences that were not previously translated into French and that now are. What it seems is that some translation has been modified, but in exchange many have been lost and now they appear in Spanish. Maybe those aren't used in LenMus.

Therefore, I have doubts about what to do with your wxWidgets_fr.po file.

As to all the suggestions and issues you posted in the "Modifications.txt" file I will study them one by one and take the necessary actions. I will inform you.

Also I would appreciate if you could agree to be contacted, from time to time, for updates in the French translation, before any new release.

If you are interested, you could join the project to do any task you would like. Not necessarily related to programming.  For instance, until now you have been the translator with more technical skills: you didn't have problems to download and build langtool or to deal directly with ebooks format. Translations management takes me time as I have to explain things to translators, answer their doubts, supervise and review the translations (of course not the texts but general issues such as formats, embedded tags, etc.), to check that the translations work and contact and coordinate all translator when in preparation for a new release. Also, there is a need think and to organize the infrastructure for translation: select and test the tools to use, prepare scripts, instructions, etc. If some one could take this responsibility it would be great! But collaboration in any other task is also welcome!

Merci encore beaucoup pour le travail accompli.
Cordialement,
Cecilio

**Comment by fred555547** [comment]
Buenas tardes Cecilio,
Es para mi un placer de dar tiempo en su projecto (hace mucho tiempo, he estudiado el espanol, y sabia hablarlo y escribirlo). Trato de tocar la guitarra (espanola), y necesito ir mas lejos : leer particiones, entenderlas, educar mi oido ,..
La traduccion me ha permitido ver la differencia entre el tiempo y las fracciones o el tiempo (una duracion) y una pulsacion (un instante) y muchas otras cosas. Y por todo eso, le agradesco.
(Que me perdona por la fotas y la ortografia)
Me llamo Frédéric Baudrand and you can contact me.
As I told you, I did the translation the dirty way (unziping files, using Atom, ziping back) to visualize both the spanish and the french text. So I had the context and I could see the result immediately lenmus. But I know it's not portable. So, this is why, I'll perform it with langtool.
I would be proud to participate to the translation of the new books, and if you agree, I'll do it from spanish. And I'm interested to see the work in progress (and so I can ask questions)
Again for wxWidgets, I used the spanish text, and started to translate the words/sentences used in the front end. Too quick and dirty. I've to work on this.
For the two books, you might validate (see and check in lenmus) them :
 I created a style to put the images in the center of the page,
I modified scores to fit the translation (some on two lines ...)
In the L2 (I didn't provide you yet because I have a lot of questions about the content) I did the following experiment in lesson 8
`<ldpmusic>
 (score (vers 1.6)(opt StaffLines.Hide yes)
  (instrument
    (musicData
      (time 2 4)
      (line (startPoint (dx -50.0) (dy 0.0)) (endPoint (dx -50.0) (dy 40.0))(width 0.5))
      (line (startPoint (dx -50.0) (dy 20.0)) (endPoint (dx 284) (dy 20.0))(width 0.5))
      (n b4 s g+)(n b4 s)(n b4 e l g-)(n b4 e g+)(n b4 s)(n b4 s g-)
      (barline)
      (n b4 s g+)(n b4 s g-)(n b4 q (stem up))(n b4 s g+)(n b4 s g-)
      (barline)
    )
  )
 )
</ldpmusic>
Initially, I tried to use a 1 line score, and it was the top line. So I tried to move up the time signature, but i does not work. This is the result. Tell me if you're interested.
`
I would be interested in joining the project, and why not to support you  in the translation management. We need to discuss this.

Hasta luego (A bientôt donc)
Cordialmente
Federico

**Comment by cecilios** [comment]
Hi Frédéric,

I cannot really write in French, it would be too tiring for me. But I can write you in Spanish instead of English, if you prefer! Better for me! You can write me in French if you prefer.

> you can contact me.

Merci beaucoup Frédéric! I will include you in the list of translators to be contacted when needed, and in the program credits. But I do not have your email address. Mine is public in my profile. Please feel free to email me for any issue you would prefer to discuss by private email.

Currently I'm really busy and cannot devote much time to LenMus and to answering emails. So please excuse me but I can not now get involved in long conversations. I expect that in two or three weeks I could have more time for this. But I will try to answer here the most urgent questions. I will answer all others in a few days.

> I did the translation the dirty way ... I'll perform it with langtool.

Do not expend time on this. I will try to preserve your translated eBooks when needing to rebuild the eBooks. It is better if we focus the effort on new coming eBooks that will replace current ones. Current roadmap is to not modify any of the existing eBooks and, instead, review the material from a pedagogical point of view, reorganize the material and rewrite new eBooks, with the idea of creating one book for each level, with all the lessons and exercises needed for that level, instead of having the material subdivided and spread in several books (theory, harmony, music reading,...) and including new exercises.

> ... the translation of the new books, and if you agree, I'll do it from Spanish

No problem, as you prefer. It is just to start with xxxx_es.po than with an empty xxxx_fr.po file, and it does not cause any problem.

> I'm interested to see the work in progress (and so I can ask questions)

We have started to work on this. A repo for discussion has been set at https://github.com/lenmus/roadmap (conversations in Spanish, not yet very focused and quite scattered). Any one is welcome to participate.

> Again for wxWidgets, I used the Spanish text ...

OK, then I will not use the wxWidgets_fr.po file you send me. If you would like to improve wxWidgets strings it is better that you contribute to the official wxWidgets repo.

> For the two books, you might validate (see and check in lenmus) them .. I created a style ... I modified scores ...

Yes, I was looking at all issues you wrote in file "Modifications.txt" and looking at the changes you did. I will agree with most of them, but I will write you later to discuss each issue.

> I would be interested in joining the project

Welcome! You can do whatever task you'd prefer! The suggestion of taking responsibility as translations manager is a possibility. I'm not happy with langtool and with using PO files for eBooks. It is perfect for translating the GUI but not for books. Perhaps other tools oriented to translating documents (and maintaining the translations) so that the translator can see the sentences in context would be better. But if you can program in C++ you can also help with many programming tasks.

When I started to develop LenMus many years ago my main problem was how to draw music scores. Years later, I extracted all the code for dealing with scores in a library, the lomse project, so that it can be reused for many other projects. See the repo at https://github.com/lenmus/lomse

> I did the following experiment in lesson 8 ...

Really, really interesting and very clever use of LDP for achieving the desired result! Congratulations!

But instead of using tricks, such as hiding the staff and drawing a line, it would be better to modify the code (and LDP) so that it would be possible to select the staff-lines to draw. This could be a good first task to do to familiarize with lenmus code and to start contributing. I will guide you in all this if you like the idea.

As to the question of creating this type of images for the eBooks when introducing a new rhythm, I like very much the idea. But as said, current roadmap is to not modify any of the existing eBooks. I will propose to the musicians your idea for the new eBooks.

A bientôt,
Cecilio

**Comment by cecilios** [comment]
Hi Frédéric,
There is a bug in the eBooks, detected while preparing the macOS package, that require to rebuild all eBooks. This will destroy you translated eBooks. How are you doing with the proper translation of eBooks?
